### PR TITLE
Thread fixes

### LINKED ISSUE #58: Race condition when accessing ThreadPool.workers
Threads are started before being assigned to the ``_workers`` dict, meaning that
if the thread then tries to access itself via ``_workers`` a race condition can
occur.

See: https://github.com/ImperialCollegeLondon/iPANACEA/issues/29

### PR DESCRIPTION
**Pull request by alexdewar**
Supersedes #59.

In the end, I had to base the changes for fixing #60 on the code in #59 anyway (as I changed how ``ThreadPool._workers`` was accessed), so I think it makes sense to have this in a single PR.

Fixes #58 and #60.

### CODE REVIEW

**Comment by alexdewar** [comment]
This is still in draft because I haven't finished testing it properly yet!

File: guikit/threads.py [hunk 1]
```diff
@@ -121,11 +121,34 @@ def __new__(cls, window: Optional[wx.Frame] = None):
             cls._instance = object.__new__(cls)
             cls._instance._window = window
             cls._instance._workers = {}
+            cls._instance._workers_lock = threading.Lock()
+
+            cls._instance._window.Bind(
```
Comment (dalonsoa) [inline guikit/threads.py:126]: This doesn't feel right. Having the `ThreadPool` controlling how and when the main window is closed seems odd. I would put this functionality in here https://github.com/ImperialCollegeLondon/guikit/blob/cb60109050463b4c9ce1e5ff89f23d2c405f9200/guikit/core.py#L211

You can bind the event in the `OnInit` method and then have a separate method (the same one you have now, but on that class) stopping all threads and destroying the main window.

File: guikit/threads.py [hunk 2]
```diff
@@ -169,7 +192,10 @@ def run_thread(
         thread = WorkerThread(target, on_abort, on_complete, on_error, daemon)
         thread.connect_events(self._window)
         thread.start()
-        self._workers[thread.ident] = thread
+
+        with self._workers_lock:
```
Comment (dalonsoa) [inline guikit/threads.py:196]: If done this way, the thread might attempt to access the workers dict before it is locked. Shouldn't be `thread.start()` within the locked block?

File: guikit/threads.py [hunk 3]
```diff
@@ -196,17 +217,51 @@ def abort_thread(self, ident: int) -> None:
...
+            workers = self._workers.values()
+
+        # Tell workers that we want to abort
+        for worker in workers:
+            worker.abort = True
+
+        # Wait for all workers to finish
+        for worker in workers:
+            worker.join()
```
Comment (dalonsoa) [inline guikit/threads.py:255]: What happens if a worker has finished after setting it to `abort` if we try to `join` it?

File: guikit/threads.py [hunk 4]
```diff
@@ -196,17 +217,51 @@ def abort_thread(self, ident: int) -> None:
...
+
+    def stop_threads_and_close_window(self):
```
Comment (dalonsoa) [inline guikit/threads.py:257]: As pointed out above, I think this would be better placed in the `MainApp`.

**Review by dalonsoa — CHANGES_REQUESTED** [review]
I've a few comments, suggestions, but it's looking good.

File: guikit/threads.py (hunk 2 above)
Comment (alexdewar) [reply guikit/threads.py:194]: Oops! Yes you're totally right. We should only release the lock once we've added the thread to ``_workers``.

File: guikit/threads.py (hunk 1 above)
Comment (alexdewar) [reply guikit/threads.py]: Good shout! Done

File: guikit/threads.py (hunk 3 above)
Comment (alexdewar) [reply guikit/threads.py:253]: That's ok too! The ``join()`` will just return instantly.

File: guikit/threads.py (hunk 4 above)
Comment (alexdewar) [reply guikit/threads.py]: :+1:

File: guikit/threads.py [hunk 5]
```diff
@@ -196,17 +215,43 @@ def abort_thread(self, ident: int) -> None:
...
+        with self._workers_lock:
+            workers = self._workers.values()
+
+        # Tell workers that we want to abort
+        for worker in workers:
+            worker.abort = True
+
+        # Wait for all workers to finish
+        for worker in workers:
+            worker.join()
```
Comment (alexdewar) [inline guikit/threads.py:253]: I've been wondering if this is a sensible thing to do:
```suggestion
            if not worker.daemon:
                worker.join()
```

That way, daemon threads will still terminate when the main window is closed, but we don't wait to close the window for e.g. iPANACEA's data acquisition thread to finish. I think this is what we want: for a GUI app it makes sense to end all the background threads when the window disappears.

What do you think?

Comment (dalonsoa) [reply guikit/threads.py:253]: It makes sense but, to be honest, it depends on what the daemon thread is doing. If it requires the main window for some reason - eg. it detects if there are updates for the software and shows a warning in the bottom left corner of the main window - then, things will fail if they are left to their own devices after the window is closed. I think it is cleaner for an app to properly deal with its threads, even if it is perfectly safe to leave them on their own.

Comment (alexdewar) [reply guikit/threads.py:253]: Okey doke. You're right that leaving it as is will be less bug prone!

**Review by dalonsoa — APPROVED** [review]

**Pull request merged**